yoyodyne/models/modules/rnn.py

```diff
@@ -304,33 +404,53 @@ def forward(
         output = self.output_proj(output)
         # Gets transition probabilities (alignment) for current states.
         alignment = self._alignment_step(decoded, encoder_out, encoder_mask)
-        return base.ModuleOutput(
-            output=output, embeddings=alignment, hiddens=hiddens
+        return base.ModuleOutput(output, hiddens, embeddings=alignment)
+
+    @property
+    def output_size(self) -> int:
+        return self.decoder_input_size + self.hidden_size
+
+
+class HardAttentionGRUDecoder(HardAttentionRNNDecoder):
+    """Zeroth-order HMM hard attention GRU decoder."""
+
+    def get_module(self) -> nn.GRU:
+        return nn.GRU(
+            self.embedding_size,
+            self.hidden_size,
+            batch_first=True,
+            bidirectional=self.bidirectional,
+            dropout=self.dropout,
+            num_layers=self.layers,
         )
 
+    @property
+    def name(self) -> str:
+        return "hard attention GRU"
+
+
+class HardAttentionLSTMDecoder(HardAttentionRNNDecoder):
+    """Zeroth-order HMM hard attention LSTM decoder."""
+
     def get_module(self) -> nn.LSTM:
         return nn.LSTM(
             self.embedding_size,
             self.hidden_size,
-            num_layers=self.layers,
-            dropout=self.dropout,
             batch_first=True,
             bidirectional=self.bidirectional,
+            dropout=self.dropout,
+            num_layers=self.layers,
         )
 
-    @property
-    def output_size(self) -> int:
-        return self.decoder_input_size + self.hidden_size
-
     @property
     def name(self) -> str:
         return "hard attention LSTM"
 
 
-class ContextHardAttentionLSTMDecoder(HardAttentionLSTMDecoder):
-    """First-order HMM hard attention LSTM."""
+class ContextualHardAttentionRNNDecoder(HardAttentionRNNDecoder):
```

Comment:
Leave as Context. (the ual can create a spelling error on a late coding debug)

Reply:
done